lib/seattleflu/id3c/cli/command/triage_uw_test_queue.py

```diff
@@ -0,0 +1,181 @@
+        - marked for surveillance testing within past 7 days
+
+    pulls primarily from REDCap instrument fields in warehouse.encounter.details, has columns like:
+        individual.identifier
+
+        redcap_record_id
+        redcap_event_name
+        redcap_repeat_instance_id
+
+        testing_determination_internal fields?
+
+        numeric priority, calculated from event instance details
+            priority calculation may be defined as a case statement or a PL/PgSQL function
+
+            priority can likely be calculated by assigning each testing
+            indication (see above) a numeric value and adding adjustments for
+            timeliness (prioritize older or newer indications?)
+
+        human-readable reason we want to test based on this event instance, for
+        internal bookkeeping/troubleshooting/viz/communicating to participant
+"""
+import click
+import logging
+from id3c.cli import cli
+from id3c.cli.command import with_database_session
+from id3c.db.session import DatabaseSession
+from ...utils import unwrap
+
+
+LOG = logging.getLogger(__name__)
+
+
+
+@cli.command("triage-uw-priority-queue", help = __doc__)
+@with_database_session
+
+@click.option("--at",
+    metavar = "<timestamp>",
+    default = "now",
+    help    = unwrap("""
+        Use quota for the given <timestamp>.
+
+        With --dry-run, can be used to test a quota schedule.  Otherwise, can
+        be used to apply unused quota from past timespans.  (Though you should
+        do that with caution, especially across day boundaries.)
+
+        Defaults to the current time."""))
+
+def triage_uw_test_queue(*, at: str, db: DatabaseSession):
+    LOG.debug(f"Triaging the UW Husky Coronavirus Testing queue with quota @ {at}")
+
+    # Lookup the quota for the current time, locking it for update at the end
+    # after we make offers.
+    quota = db.fetch_row("""
+        select
+            name,
+            timespan,
+            "limit",
+            used,
+            "limit" - used as remaining
+        from
+            operations.test_quota
+        where
+            name = 'uw' and timespan @> timestamp with time zone ?
+        for update
+        """, (at,))
+
+    if not quota:
+        LOG.warning(f"No quota row found, aborting")
+        return
+
+    if not quota.remaining > 0:
+        LOG.warning(f"No quota remaining for {quota.name} during {quota.timespan}, aborting")
+        return
+
+    LOG.info(
+        f"Quota remaining for {quota.name} during {quota.timespan} "
+        f"is {quota.remaining}/{quota.limit} (remaining/limit)")
+
+    # Offer testing to the top entries in our priority queue by updating REDCap.
+    offers = db.fetch_rows("""
+        select
+            redcap_record_id,
+            redcap_event_name,
+            redcap_repeat_instance_id,
+            reason,
+            priority
+        from
+            shipping.uw_priority_queue_v1
+        limit
```

Comment:
order by priority ?

Reply:
Yes, that's important! But the intent was for `shipping.uw_priority_queue_v1` to already be ordered by priority, and that's how Jover implemented it in #151. So I think the above is good as-is.

Reply:
Since priority is key it would be safer to specify the order by here even if the view does it.

Reply:
The idea was that this command would treat the queue as a black box. It just pulls things off the top N at a time, and whatever the queue says is on top is on top. If we don't treat the queue as a black box, then when we update the queue later to break ties (which I think we need to), then we're going to have to update this command too.

But if you think the balance favors an explicit ordering here, then fair enough. I checked and Pg's planner does appear smart enough to realize the view is already sorted by the requested column and not resort.

Reply:
([Further comments](https://github.com/seattleflu/id3c-customizations/pull/151#discussion_r504165879) on the tie breaking we'd have to duplicate here if we think an explicit ordering here is advised.)

Reply:
> The idea was that this command would treat the queue as a black box. It just pulls things off the top N at a time, and whatever the queue says is on top is on top. If we don't treat the queue as a black box, then when we update the queue later to break ties (which I think we need to), then we're going to have to update this command too.
> 
> But if you think the balance favors an explicit ordering here, then fair enough. I checked and Pg's planner does appear smart enough to realize the view is already sorted by the requested column and not resort.

I'll drop this. It's not a big deal as long as the data gets to the code sorted.